### PR TITLE
hardening/562_batching_timeout_and_transaction_ids_accumulation

### PR DESCRIPTION
**Pull request by frbattid**
* Implements issue #562 
* 100% unit tests passed:
```
Tests run: 59, Failures: 0, Errors: 0, Skipped: 0
```
* (unofficial) e2e tests passed:
```
---------------------------------------
timeout reached for an incomplete batch
---------------------------------------
time=2015-10-29T14:46:14.596CET | lvl=INFO | trans=1446126354-390-0000000010 | function=process | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionSink[246] : Batch accumulation time reached, the batch will be processed as it is
time=2015-10-29T14:46:14.600CET | lvl=INFO | trans=1446126354-390-0000000010 | function=persistAggregation | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionHDFSSink[850] : [hdfs-sink] Persisting data at OrionHDFSSink. HDFS file (def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room/room1_room.txt), Data ({"recvTimeTs":"1446126356","recvTime":"2015-10-29T13:45:56.73Z","fiware-servicepath":"def_serv_path_deleteme_1","entityId":"Room1","entityType":"Room","attrName":"temperature","attrType":"centigrade","attrValue":"0","attrMd":[{"name":"resolution","type":"integer","value":"0"}]}
{"recvTimeTs":"1446126356","recvTime":"2015-10-29T13:45:56.170Z","fiware-servicepath":"def_serv_path_deleteme_1","entityId":"Room1","entityType":"Room","attrName":"temperature","attrType":"centigrade","attrValue":"1","attrMd":[{"name":"resolution","type":"integer","value":".50"}]}
...
{"recvTimeTs":"1446126356","recvTime":"2015-10-29T13:45:56.527Z","fiware-servicepath":"def_serv_path_deleteme_1","entityId":"Room1","entityType":"Room","attrName":"temperature","attrType":"centigrade","attrValue":"9","attrMd":[{"name":"resolution","type":"integer","value":"4.50"}]}
{"recvTimeTs":"1446126356","recvTime":"2015-10-29T13:45:56.573Z","fiware-servicepath":"def_serv_path_deleteme_1","entityId":"Room1","entityType":"Room","attrName":"temperature","attrType":"centigrade","attrValue":"10","attrMd":[{"name":"resolution","type":"integer","value":"5.00"}]})
time=2015-10-29T14:46:15.499CET | lvl=INFO | trans=1446126354-390-0000000010 | function=provisionHiveTable | comp=Cygnus | msg=com.telefonica.iot.cygnus.backends.hdfs.HDFSBackendImplREST[201] : Creating Hive external table=frb_def_serv_deleteme_1_def_serv_path_deleteme_1_room1_room_row
time=2015-10-29T14:46:16.417CET | lvl=INFO | trans=1446126354-390-0000000010 | function=process | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionSink[295] : Finishing transaction (1446126354-390-0000000000,1446126354-390-0000000001,1446126354-390-0000000002,1446126354-390-0000000003,1446126354-390-0000000004,1446126354-390-0000000005,1446126354-390-0000000006,1446126354-390-0000000007,1446126354-390-0000000008,1446126354-390-0000000009,1446126354-390-0000000010)
time=2015-10-29T14:46:38.437CET | lvl=INFO | trans=1446126354-390-0000000010 | function=process | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionSink[246] : Batch accumulation time reached, the batch will be processed as it is
time=2015-10-29T14:46:38.438CET | lvl=INFO | trans=1446126354-390-0000000010 | function=process | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionSink[295] : Finishing transaction ()
----------------------------------
timeout reached for an empty batch
----------------------------------
time=2015-10-29T14:47:00.462CET | lvl=INFO | trans=1446126354-390-0000000010 | function=process | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionSink[246] : Batch accumulation time reached, the batch will be processed as it is
time=2015-10-29T14:47:00.463CET | lvl=INFO | trans=1446126354-390-0000000010 | function=process | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionSink[295] : Finishing transaction ()
----------------------------
batch completed (no timeout)
----------------------------
time=2015-10-29T14:48:53.775CET | lvl=INFO | trans=1446126354-390-0000000110 | function=persistAggregation | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionHDFSSink[850] : [hdfs-sink] Persisting data at OrionHDFSSink. HDFS file (def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room/room1_room.txt), Data ({"recvTimeTs":"1446126529","recvTime":"2015-10-29T13:48:49.271Z","fiware-servicepath":"def_serv_path_deleteme_1","entityId":"Room1","entityType":"Room","attrName":"temperature","attrType":"centigrade","attrValue":"0","attrMd":[{"name":"resolution","type":"integer","value":"0"}]}
{"recvTimeTs":"1446126529","recvTime":"2015-10-29T13:48:49.321Z","fiware-servicepath":"def_serv_path_deleteme_1","entityId":"Room1","entityType":"Room","attrName":"temperature","attrType":"centigrade","attrValue":"1","attrMd":[{"name":"resolution","type":"integer","value":".50"}]}
...
{"recvTimeTs":"1446126533","recvTime":"2015-10-29T13:48:53.712Z","fiware-servicepath":"def_serv_path_deleteme_1","entityId":"Room1","entityType":"Room","attrName":"temperature","attrType":"centigrade","attrValue":"98","attrMd":[{"name":"resolution","type":"integer","value":"49.00"}]}
{"recvTimeTs":"1446126533","recvTime":"2015-10-29T13:48:53.751Z","fiware-servicepath":"def_serv_path_deleteme_1","entityId":"Room1","entityType":"Room","attrName":"temperature","attrType":"centigrade","attrValue":"99","attrMd":[{"name":"resolution","type":"integer","value":"49.50"}]})
time=2015-10-29T14:48:55.372CET | lvl=INFO | trans=1446126354-390-0000000110 | function=process | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionSink[295] : Finishing transaction (1446126354-390-0000000011,1446126354-390-0000000012,1446126354-390-0000000013,1446126354-390-0000000014,1446126354-390-0000000015,1446126354-390-0000000016,1446126354-390-0000000017,1446126354-390-0000000018,1446126354-390-0000000019,1446126354-390-0000000020,1446126354-390-0000000021,1446126354-390-0000000022,1446126354-390-0000000023,1446126354-390-0000000024,1446126354-390-0000000025,1446126354-390-0000000026,1446126354-390-0000000027,1446126354-390-0000000028,1446126354-390-0000000029,1446126354-390-0000000030,1446126354-390-0000000031,1446126354-390-0000000032,1446126354-390-0000000033,1446126354-390-0000000034,1446126354-390-0000000035,1446126354-390-0000000036,1446126354-390-0000000037,1446126354-390-0000000038,1446126354-390-0000000039,1446126354-390-0000000040,1446126354-390-0000000041,1446126354-390-0000000042,1446126354-390-0000000043,1446126354-390-0000000044,1446126354-390-0000000045,1446126354-390-0000000046,1446126354-390-0000000047,1446126354-390-0000000048,1446126354-390-0000000049,1446126354-390-0000000050,1446126354-390-0000000051,1446126354-390-0000000052,1446126354-390-0000000053,1446126354-390-0000000054,1446126354-390-0000000055,1446126354-390-0000000056,1446126354-390-0000000057,1446126354-390-0000000058,1446126354-390-0000000059,1446126354-390-0000000060,1446126354-390-0000000061,1446126354-390-0000000062,1446126354-390-0000000063,1446126354-390-0000000064,1446126354-390-0000000065,1446126354-390-0000000066,1446126354-390-0000000067,1446126354-390-0000000068,1446126354-390-0000000069,1446126354-390-0000000070,1446126354-390-0000000071,1446126354-390-0000000072,1446126354-390-0000000073,1446126354-390-0000000074,1446126354-390-0000000075,1446126354-390-0000000076,1446126354-390-0000000077,1446126354-390-0000000078,1446126354-390-0000000079,1446126354-390-0000000080,1446126354-390-0000000081,1446126354-390-0000000082,1446126354-390-0000000083,1446126354-390-0000000084,1446126354-390-0000000085,1446126354-390-0000000086,1446126354-390-0000000087,1446126354-390-0000000088,1446126354-390-0000000089,1446126354-390-0000000090,1446126354-390-0000000091,1446126354-390-0000000092,1446126354-390-0000000093,1446126354-390-0000000094,1446126354-390-0000000095,1446126354-390-0000000096,1446126354-390-0000000097,1446126354-390-0000000098,1446126354-390-0000000099,1446126354-390-0000000100,1446126354-390-0000000101,1446126354-390-0000000102,1446126354-390-0000000103,1446126354-390-0000000104,1446126354-390-0000000105,1446126354-390-0000000106,1446126354-390-0000000107,1446126354-390-0000000108,1446126354-390-0000000109,1446126354-390-0000000110)
```
* Assignee @fgalan

### CODE REVIEW

File: doc/design/OrionHDFSSink.md [hunk 1]
```diff
@@ -260,12 +262,14 @@ There exists an [issue](https://github.com/telefonicaid/fiware-cosmos/issues/111
...
 
-By default, `OrionHDFSSink` has a configured batch size of one, but as explained above, it is highly recommended to increase the batch size for performance purposes. Which is the optimal value? The size of the batch it is closely related to the transaction size of the channel the events are got from (it has no sense the first one is greater then the second one), and it depends on the number of estimated sub-batches as well. A deeper discussion on the batches of events and their appropriate sizing may be found in the [performance document](../operation/performance_tuning_tips.md).
+The batch mechanism adds an accumlation timeout to protect the sink to stay in an eternal state of batch building when no new data arrives. If such a timeout is reached, then the batch is persisted as it is.
```
Comment (fgalan) [inline doc/design/OrionHDFSSink.md]: typo: accumlation

File: doc/design/OrionHDFSSink.md [hunk 2]
```diff
@@ -260,12 +262,14 @@ There exists an [issue](https://github.com/telefonicaid/fiware-cosmos/issues/111
...
+
+By default, `OrionHDFSSink` has a configured batch size and batch accumulation timeout of 1 and 30 seconds, respectively. Nevertheless, as explained above, it is highly recommended to increase at least the batch size for performance purposes. Which are the optimal values? The size of the batch it is closely related to the transaction size of the channel the events are got from (it has no sense the first one is greater then the second one), and it depends on the number of estimated sub-batches as well. The accumulation timeout will depend on how often you want to see new data in the final storage. A deeper discussion on the batches of events and their appropriate sizing may be found in the [performance document](../operation/performance_tuning_tips.md).
```
Comment (fgalan) [inline doc/design/OrionHDFSSink.md]: Idea for the future: adaptatitive accumulation time.

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionSink.java [hunk 3]
```diff
@@ -245,13 +229,74 @@ public Status process() throws EventDeliveryException {
...
                 } // if else
-            } finally {
-                LOGGER.info("Finishing transaction (" + MDC.get(Constants.HEADER_TRANSACTION_ID) + ")");
-            } // try catch finally
+            } // try catch
         } else {
+            // get and process as many events as the batch size
+            int currentIndex;
+            
+            for (currentIndex = accumulator.getAccIndex(); currentIndex < batchSize; currentIndex++) {
```
Comment (fgalan) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionSink.java]: So you check the accumulation timeout each time a new event arrives to the buffer, right?

Comment (frbattid) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionSink.java]: Not exactly. I check for the timeout each time I'm going to read an event. The `for` loop does not ensure an event is read, it is only an attempt. If there are no events within the channel, then the `process()` method is exited; later, a new `process()` invocation will be done by Flume core. I admit by exiting and re-runing the `process()` method the timeouts are not triggered at exact time intervals, but the final behaviour is very close to that.

File: doc/design/OrionHDFSSink.md (hunk 2 above)
Comment (frbattid) [inline doc/design/OrionHDFSSink.md]: Nice: https://github.com/telefonicaid/fiware-cygnus/issues/583

File: doc/design/OrionHDFSSink.md (hunk 1 above)
Comment (frbattid) [inline doc/design/OrionHDFSSink.md]: Fixed in https://github.com/telefonicaid/fiware-cygnus/commit/048eecb2843567c6788449fb8c10bbc6773450e5

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionSink.java (hunk 3 above)
Comment (fgalan) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionSink.java]: NTC

**Comment by fgalan** [comment]
LGTM

**Pull request merged**